test/test-ipc-send-recv.c

```diff
@@ -189,48 +242,93 @@ static void read_cb(uv_stream_t* handle,
-    r = uv_pipe_init(ctx.channel.loop, &ctx.recv.pipe, 0);
+    r = uv_pipe_init(ctx2.channel.loop, &ctx2.recv.pipe, 0);
   else if (pending == UV_TCP)
-    r = uv_tcp_init(ctx.channel.loop, &ctx.recv.tcp);
+    r = uv_tcp_init(ctx2.channel.loop, &ctx2.recv.tcp);
   else
     abort();
   ASSERT(r == 0);
 
-  r = uv_accept(handle, &ctx.recv.stream);
+  r = uv_accept(handle, &ctx2.recv.stream);
   ASSERT(r == 0);
 
-  r = uv_write2(&ctx.write_req,
-                (uv_stream_t*)&ctx.channel,
+  r = uv_write2(&ctx2.write_req,
+                (uv_stream_t*)&ctx2.channel,
                 &wrbuf,
                 1,
-                &ctx.recv.stream,
+                &ctx2.recv.stream,
                 write2_cb);
   ASSERT(r == 0);
 }
 
-
-/* stdin is a duplex channel over which a handle is sent.
- * We receive it and send it back where it came from.
- */
-int ipc_send_recv_helper(void) {
+static void send_recv_start() {
   int r;
+  ASSERT(1 == uv_is_readable((uv_stream_t*)&ctx2.channel));
+  ASSERT(1 == uv_is_writable((uv_stream_t*)&ctx2.channel));
+  ASSERT(0 == uv_is_closing((uv_handle_t*)&ctx2.channel));
 
-  memset(&ctx, 0, sizeof(ctx));
+  r = uv_read_start((uv_stream_t*)&ctx2.channel, alloc_cb, read_cb);
+  ASSERT(r == 0);
+}
+
+static void listen_cb(uv_stream_t* handle, int status) {
+  int r;
+  ASSERT(handle == (uv_stream_t*)&ctx2.listen);
+  ASSERT(status == 0);
 
-  r = uv_pipe_init(uv_default_loop(), &ctx.channel, 1);
+  r = uv_accept((uv_stream_t*)&ctx2.listen, (uv_stream_t*)&ctx2.channel);
   ASSERT(r == 0);
 
-  uv_pipe_open(&ctx.channel, 0);
-  ASSERT(1 == uv_is_readable((uv_stream_t*)&ctx.channel));
-  ASSERT(1 == uv_is_writable((uv_stream_t*)&ctx.channel));
-  ASSERT(0 == uv_is_closing((uv_handle_t*)&ctx.channel));
+  send_recv_start();
+}
+
+int run_ipc_send_recv_helper(uv_loop_t* loop, int inprocess) {
+  int r;
+
+  memset(&ctx2, 0, sizeof(ctx2));
 
-  r = uv_read_start((uv_stream_t*)&ctx.channel, alloc_cb, read_cb);
+  r = uv_pipe_init(loop, &ctx2.listen, 0);
   ASSERT(r == 0);
 
-  r = uv_run(uv_default_loop(), UV_RUN_DEFAULT);
+  r = uv_pipe_init(loop, &ctx2.channel, 1);
   ASSERT(r == 0);
 
-  MAKE_VALGRIND_HAPPY();
+  if (inprocess) {
+    r = uv_pipe_bind(&ctx2.listen, TEST_PIPENAME_2);
+    ASSERT(r == 0);
+
+    r = uv_listen((uv_stream_t*)&ctx2.listen, SOMAXCONN, listen_cb);
+    ASSERT(r == 0);
+  } else {
+    r = uv_pipe_open(&ctx2.channel, 0);
+    ASSERT(r == 0);
+
+    send_recv_start();
+  }
+
+  r = uv_run(loop, UV_RUN_DEFAULT);
+  ASSERT(r == 0);
+
+  /* Same work as MAKE_VALGRIND_HAPPY */
+  close_loop(loop);
+  uv_loop_delete(loop);
   return 0;
 }
+
+/* stdin is a duplex channel over which a handle is sent.
+ * We receive it and send it back where it came from.
+ */
+int ipc_send_recv_helper(void) {
+  return run_ipc_send_recv_helper(uv_default_loop(), FALSE);
+}
+
+void ipc_send_recv_helper_threadproc(void* arg) {
+  int r;
+  uv_loop_t* loop = (uv_loop_t*)malloc(sizeof(uv_loop_t));
```

Comment:
I did at first, but the two steps from MAKE_VALGRIND_HAPPY tried to free() it. Should I change to a stack variable and also remove the call to uv_loop_delete(loop)?